Add bitmap control for perf context

LemonHX opened the pull request:
[Design document](https://docs.google.com/document/d/1JYmWMIZwYV0AZW6rNv_oFVZLBCAkxLLYslt39eEZstM/edit?usp=sharing)

[related issue No.543 in rust-rocksdb](https://github.com/tikv/rust-rocksdb/issues/543)

LemonHX committed:
`PerfFlag` implementation aside the `PerfLevel`
design spec in on: https://docs.google.com/document/d/1JYmWMIZwYV0AZW6rNv_oFVZLBCAkxLLYslt39eEZstM/edit?usp=sharing

Signed-off-by: lemonhx <lemonhx@lemonhx.tech>

Connor1996 commented on include/rocksdb/c.h:
```diff
@@ -1119,7 +1119,11 @@ enum {
   rocksdb_total_metric_count = 68
 };
 
+#include "perf_flag_defs.h"
```
is it necessary?

LemonHX replied:
yes, but not that much, It can be removed but in perf level is also included, thats whay I included.

Connor1996 reviewed (commented):
LGTM overall

LemonHX committed:
Adding some test and fixing git errors
fix indentation error in `c.cc`
adding comment on `FLAG_END`
explain what does `& 0b111` does
fix a huge mistake in `CheckPerfFlag`
and writing the edge case tests for `perf_flag`

Signed-off-by: lemonhx <lemonhx@lemonhx.tech>

breeswish commented on monitoring/perf_flag.cc:
```diff
@@ -0,0 +1,29 @@
+void EnablePerfFlag(uint64_t flag) {
+  if (CheckPerfFlag(flag)) {
+  } else {
+    GET_FLAG(flag) ^= (uint64_t)0b1 << ((uint64_t)flag & (uint64_t)0b111);
```
Why do we need this part? `& (uint64_t)0b111`

LemonHX replied:
& 0b111 means find where is the flag located
which is equivalent to % 8 and faster

LemonHX commented on include/rocksdb/perf_flag_defs.h:
```diff
@@ -0,0 +1,107 @@
+  flag_logger_nanos,
+  flag_cpu_read_nanos,
+  flag_cpu_write_nanos,
+  FLAG_END
```
I'll do that

breeswish commented on monitoring/perf_flag_test.cc:
```diff
@@ -0,0 +1,82 @@
+        SetPerfLevel(PerfLevel::kDisable);
+        EnablePerfFlag(flag_user_key_comparison_count);
+      },
+      { ASSERT_GT(get_perf_context()->user_key_comparison_count, 0); });
```
Perf context is stored locally so that it may contain data from previous tests. I would suggest you to assert the delta value.

yiwu-arbug commented:
/run-tests

yiwu-arbug commented:
@LemonHX build fail. along with CMakeList.txt, you also need to add the new files and new tests to src.mk and https://github.com/tikv/rocksdb/blob/6.4.tikv/Makefile#L420

yiwu-arbug commented:
/run-tests

LemonHX committed:
add source to `src.mk`
add test into `Makefile`
and removed unused variable.
fix indentation error in `CMakeLists`

Signed-off-by: lemonhx <lemonhx@lemonhx.tech>

LemonHX commented:
> 
> 
> @LemonHX build fail. along with CMakeList.txt, you also need to add the new files and new tests to src.mk and https://github.com/tikv/rocksdb/blob/6.4.tikv/Makefile#L420

done

LemonHX commented:
/run-tests

yiwu-arbug commented:
/run-tests

yiwu-arbug commented on CMakeLists.txt:
```diff
@@ -974,7 +975,8 @@ if(WITH_TESTS)
         memtable/write_buffer_manager_test.cc
         monitoring/histogram_test.cc
-        monitoring/iostats_context_test.cc
+          monitoring/perf_flag_test.cc
```
indent

tabokie replied:
Please revert indentation changes.

yiwu-arbug commented:
The encryption_env test and Mac test in CI are broken. We can ignore them for now.

LemonHX commented:
so could we merge it?

tabokie commented on monitoring/perf_flag_test.cc:
```diff
@@ -0,0 +1,82 @@
+int main(int argc, char** argv) {
+  ::testing::InitGoogleTest(&argc, argv);
+  return RUN_ALL_TESTS();
+}
```
Need a newline

tabokie commented on monitoring/perf_flag_test.cc:
```diff
@@ -0,0 +1,82 @@
+  }
+
+TEST(PerfFlagTest, TestEnableFlagStandAlone) {
+  DB_TEST_HELPER(
```
Macro is unnecessary, you can refactor some of the code into a `Prepare` member function in `PerfFlagTest`.

tabokie commented on monitoring/perf_flag_test.cc:
```diff
@@ -0,0 +1,82 @@
+#include "rocksdb/slice.h"
+
+namespace rocksdb {
+class PerfFlagTest : testing::Test {};
+TEST(PerfFlagTest, TestEnableFlag) {
+  for (int i = 0; i < 10; ++i) {
+    EnablePerfFlag(i);
+    ASSERT_EQ(CheckPerfFlag(i), true);
+  }
+}
+TEST(PerfFlagTest, TestDisableFlag) {
+  for (int i = 0; i < 10; ++i) {
+    EnablePerfFlag(i);
+  }
+  for (int i = 0; i < 10; ++i) {
+    DisablePerfFlag(i);
+    ASSERT_EQ(CheckPerfFlag(i), false);
+  }
+}
```
```suggestion
class PerfFlagTest : testing::Test {};

TEST(PerfFlagTest, TestEnableFlag) {
  for (int i = 0; i < 10; ++i) {
    EnablePerfFlag(i);
    ASSERT_EQ(CheckPerfFlag(i), true);
  }
}

TEST(PerfFlagTest, TestDisableFlag) {
  for (int i = 0; i < 10; ++i) {
    EnablePerfFlag(i);
  }
  for (int i = 0; i < 10; ++i) {
    DisablePerfFlag(i);
    ASSERT_EQ(CheckPerfFlag(i), false);
  }
}
```

tabokie commented on monitoring/perf_flag.cc:
```diff
@@ -0,0 +1,30 @@
+  } else {
+    // & 0b111 means find the flag location is a alternative way to do mod operation
+    GET_FLAG(flag) ^= (uint64_t)0b1 << ((uint64_t)flag & (uint64_t)0b111);
+  }
+}
+
+void DisablePerfFlag(uint64_t flag) {
+  if (CheckPerfFlag(flag)) {
+    GET_FLAG(flag) ^= (uint64_t)0b1 << ((uint64_t)flag & (uint64_t)0b111);
+  } else {
+  }
+}
+
+bool CheckPerfFlag(uint64_t flag) {
+  return ((uint64_t)GET_FLAG(flag) & (uint64_t)0b1
+                                         << (flag & (uint64_t)0b111)) != 0;
+}
+
+}  // namespace rocksdb
```
newline

tabokie commented on monitoring/perf_flag.cc:
```diff
@@ -0,0 +1,30 @@
+void DisablePerfFlag(uint64_t flag) {
+  if (CheckPerfFlag(flag)) {
+    GET_FLAG(flag) ^= (uint64_t)0b1 << ((uint64_t)flag & (uint64_t)0b111);
+  } else {
+  }
```
What is this for?

LemonHX replied:
for doing nothing

tabokie commented on Makefile:
```diff
@@ -418,6 +418,7 @@ ANALYZETOOLOBJECTS = $(ANALYZER_LIB_SOURCES:.cc=.o)
 EXPOBJECTS = $(LIBOBJECTS) $(TESTUTIL)
 
 TESTS = \
+	perf_flag_test \
```
Share we put it beside `perf_context_test`? Same goes for the entry item below.

tabokie commented on monitoring/perf_flag.cc:
```diff
@@ -0,0 +1,30 @@
+#endif
+
+void EnablePerfFlag(uint64_t flag) {
+  if (CheckPerfFlag(flag)) {
+  } else {
```
What is this for?

tabokie commented on monitoring/perf_context_imp.h:
```diff
@@ -38,24 +38,24 @@ extern thread_local PerfContext perf_context;
 // Declare and set start time of the timer
 #define PERF_TIMER_GUARD(metric)                                  \
-  PerfStepTimer perf_step_timer_##metric(&(perf_context.metric)); \
+  PerfStepTimer perf_step_timer_##metric(&(perf_context.metric),CheckPerfFlag(flag_##metric)); \
```
```suggestion
  PerfStepTimer perf_step_timer_##metric(&(perf_context.metric), CheckPerfFlag(flag_##metric)); \
```

tabokie replied:
And could you also align the `\`? thanks

tabokie commented on monitoring/perf_context_imp.h:
```diff
@@ -38,24 +38,24 @@ extern thread_local PerfContext perf_context;
 // Declare and set start time of the timer
 #define PERF_TIMER_GUARD_WITH_ENV(metric, env)                         \
-  PerfStepTimer perf_step_timer_##metric(&(perf_context.metric), env); \
+  PerfStepTimer perf_step_timer_##metric(&(perf_context.metric),CheckPerfFlag(flag_##metric),env); \
```
ditto

tabokie commented on monitoring/perf_context_imp.h:
```diff
@@ -38,24 +38,24 @@ extern thread_local PerfContext perf_context;
 #define PERF_CPU_TIMER_GUARD(metric, env)              \
   PerfStepTimer perf_step_timer_##metric(              \
-      &(perf_context.metric), env, true,               \
+      &(perf_context.metric),CheckPerfFlag(flag_##metric), env, true,               \
```
ditto

tabokie commented on monitoring/perf_context_imp.h:
```diff
@@ -38,24 +38,24 @@ extern thread_local PerfContext perf_context;
 #define PERF_CONDITIONAL_TIMER_FOR_MUTEX_GUARD(metric, condition, stats,       \
                                                ticker_type)                    \
-  PerfStepTimer perf_step_timer_##metric(&(perf_context.metric), nullptr,      \
+  PerfStepTimer perf_step_timer_##metric(&(perf_context.metric),CheckPerfFlag(flag_##metric), nullptr,      \
```
ditto

tabokie commented on include/rocksdb/perf_flag.h:
```diff
@@ -0,0 +1,15 @@
+  (((uint64_t)FLAG_END & (uint64_t)0b111) == 0  \
+       ? ((uint64_t)FLAG_END >> (uint64_t)0b11) \
+       : ((uint64_t)FLAG_END >> (uint64_t)0b11) + (uint64_t)1)
+namespace rocksdb{
+  void EnablePerfFlag(uint64_t flag);
+  void DisablePerfFlag(uint64_t flag);
+  bool CheckPerfFlag(uint64_t flag);
+}
```
```suggestion
namespace rocksdb {
  void EnablePerfFlag(uint64_t flag);
  void DisablePerfFlag(uint64_t flag);
  bool CheckPerfFlag(uint64_t flag);
} // namespace rocksdb
```

tabokie commented on include/rocksdb/perf_flag.h:
```diff
@@ -0,0 +1,15 @@
+// complements to perf_level
+#include <cstdint>
```
Header files should be put below `#pragma once` as well.

tabokie commented on db/c.cc:
```diff
@@ -2749,6 +2753,18 @@ void rocksdb_set_perf_level(int v) {
   SetPerfLevel(level);
 }
 
+void rocksdb_enable_perf_flag(uint64_t flag){
+  EnablePerfFlag(flag);
+}
+
+void rocksdb_disable_perf_flag(uint64_t flag){
+  DisablePerfFlag(flag);
+}
+
+int rocksdb_check_perf_flag(uint64_t flag){
+  return (int)CheckPerfFlag(flag);
+}
```
```suggestion
void rocksdb_enable_perf_flag(uint64_t flag) {
  EnablePerfFlag(flag);
}

void rocksdb_disable_perf_flag(uint64_t flag) {
  DisablePerfFlag(flag);
}

int rocksdb_check_perf_flag(uint64_t flag) {
  return static_cast<int>(CheckPerfFlag(flag));
}
```

tabokie commented:
@yiwu-arbug Are we running clang-format in CI right now? I wonder how these styling issues go unalerted.

LemonHX committed:
fix formatting and styling issues.

Co-authored-by: Xinye Tao <xy.tao@outlook.com>
Signed-off-by: lemonhx <lemonhx@lemonhx.tech>

tabokie commented:
@LemonHX Could you run this line in your repo to format this PR manually? I'm not sure if there's any issues I haven't sighted.
```
git diff -U0 f95aaf8dac667334bb6314bd06d80749f63177bb | python <(curl https://raw.githubusercontent.com/llvm-mirror/clang/master/tools/clang-format/clang-format-diff.py) -p1 -i
```
And a few commit signoffs need fixing.

tabokie committed:
address comments

Signed-off-by: tabokie <xy.tao@outlook.com>

tabokie committed:
remove redundant line

Signed-off-by: tabokie <xy.tao@outlook.com>

tabokie committed:
trigger rebuild

Signed-off-by: tabokie <xy.tao@outlook.com>

tabokie reviewed (approved):
I've manually resolved all the review comments. LGTM

Connor1996 commented on include/rocksdb/perf_flag.h:
```diff
@@ -0,0 +1,19 @@
+
+#define FLAGS_LEN                               \
+  (((uint64_t)FLAG_END & (uint64_t)0b111) == 0  \
+       ? ((uint64_t)FLAG_END >> (uint64_t)0b11) \
```
```suggestion
       ? ((uint64_t)FLAG_END >> (uint64_t)3) \
```

Connor1996 commented on include/rocksdb/perf_flag.h:
```diff
@@ -0,0 +1,19 @@
+#define FLAGS_LEN                               \
+  (((uint64_t)FLAG_END & (uint64_t)0b111) == 0  \
+       ? ((uint64_t)FLAG_END >> (uint64_t)0b11) \
+       : ((uint64_t)FLAG_END >> (uint64_t)0b11) + (uint64_t)1)
```
```suggestion
       : ((uint64_t)FLAG_END >> (uint64_t)3) + (uint64_t)1)
```

tabokie committed:
address comments

Signed-off-by: tabokie <xy.tao@outlook.com>

Connor1996 reviewed (approved):
LGTM
